### PR TITLE
Use a decorator to register features

### PR DESCRIPTION
**Pull request by mgeplf**
* This makes it easier to add features, and have aliases for them
* Will make it easier to add descriptions of the return types, in
  the future

### CODE REVIEW

File: neurom/core/features.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+from .types import tree_type_checker
+from . import iter_neurites
+from . import NeuriteType
+
+
+FEATURES = {'NEURITEFEATURES': {},
+            'NEURONFEATURES': {}, }
+
+
+def register_neurite_feature(name, func):
```
Comment (arsenius7) [inline neurom/core/features.py]: Can't see where this one is used.

File: neurom/core/features.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        raise NeuroMError('Attempt to hide registered feature %s' % name)
+
+    FEATURES[namespace][name] = func
+
+
+def feature(func=None, namespace=None, names=''):
+    '''feature decorator to automatically register the feature in the appropriate namespace'''
+
+    if func is None:
+        return partial(feature, namespace=namespace, names=names)
```
Comment (arsenius7) [inline neurom/core/features.py]: what is that?

File: neurom/core/features.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        name(string): name of the feature, used for access via get() function.
+        func(callable): single parameter function of a neurite.
+    '''
+    if name in FEATURES[namespace]:
+        raise NeuroMError('Attempt to hide registered feature %s' % name)
+
+    FEATURES[namespace][name] = func
+
+
+def feature(func=None, namespace=None, names=''):
```
Comment (arsenius7) [inline neurom/core/features.py]: parameter name 'aliases' would convey the meaning better.
And why is it a space-separated string rather than list?

File: neurom/core/features.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+
+def _indent(string, count):
+    '''indent `string` by `count` * INDENT'''
+    indent = _INDENT * count
+    ret = indent + string.replace('\n', '\n' + indent)
+    return ret.rstrip()
+
+
+def get_doc():
```
Comment (arsenius7) [inline neurom/core/features.py]: That's a lot of manual formatting going on.
I would look for a way to dump it to some hierarhy, and let pretty print yaml or whatever deal with indents.

File: neurom/core/features.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    Returns:
+        features as a 1D or 2D numpy array.
+    '''
+
+    for name in resolution_order:
+        if feature_name in FEATURES[name]:
+            feat = FEATURES[name][feature_name]
+            break
+    else:
+        raise NeuroMError('Unable to find feature: %s' % feature_name)
```
Comment (eleftherioszisis) [inline neurom/core/features.py]: If a feature exists in both resolution orders, then this would always return the ```NEURITEFEATURES``` one, right?
Why not checking the type of the object and looking up for the corresponding resolution order?

Comment (mgeplf) [inline neurom/core/features.py]: @eleftherioszisis this is so the behavior hasn't changed from the old version

File: neurom/core/features.py (hunk 1 above)
Comment (mgeplf) [inline neurom/core/features.py]: it's user feature that was added here: #537

File: neurom/core/features.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+import numpy as np
+
+from functools import partial
+from ..exceptions import NeuroMError
+from .types import tree_type_checker
+from . import iter_neurites
+from . import NeuriteType
+
+
+FEATURES = {'NEURITEFEATURES': {},
```
Comment (mgeplf) [inline neurom/core/features.py]: yeah, good point

File: neurom/core/features.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    '''
+    def _fun(neurites, neurite_type=NeuriteType.all):
+        '''Wrap neurite function from outer scope and map into list'''
+        return list(func(n)
+                    for n in iter_neurites(neurites, filt=tree_type_checker(neurite_type)))
+
+    register_feature('NEURITEFEATURES', name, _fun)
+
+
+def register_feature(namespace, name, func):
```
Comment (mgeplf) [inline neurom/core/features.py]: correct, I could keep a list of namespaces, and assert against that

File: neurom/core/features.py (hunk 3 above)
Comment (mgeplf) [inline neurom/core/features.py]: good points, will change to that

File: neurom/core/features.py (hunk 2 above)
Comment (mgeplf) [inline neurom/core/features.py]: trick to deal w/ the @feature vs. @feature() syntax

File: neurom/core/features.py (hunk 5 above)
Comment (mgeplf) [inline neurom/core/features.py]: hmm, I like it due to not needing `feature_found` variable

File: neurom/core/features.py (hunk 4 above)
Comment (mgeplf) [inline neurom/core/features.py]: I think that would be good for another commit

File: neurom/core/features.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    '''
+
+    for name in resolution_order:
+        if feature_name in FEATURES[name]:
+            feat = FEATURES[name][feature_name]
+            break
+    else:
+        raise NeuroMError('Unable to find feature: %s' % feature_name)
+
+    return np.array(list(feat(obj, **kwargs)))
```
Comment (eleftherioszisis) [inline neurom/core/features.py]: Well, provided that all the generators yield elements that can map to 1d arrays (i.e. no single numbers or nested elements), then I think it should be fine.

File: neurom/core/features.py (hunk 5 above)
Comment (eleftherioszisis) [inline neurom/core/features.py]: My concern is that this will be wrong for features that apply to both resolution orders, such as the ```principal_direction_extents``` that I suggested in #567.

Unless the user passes ```NEURONFEATURES``` as the ```resolution_order``` or swap the elements in the tuple, the feature will be assumed as a neurite feature and treat the cell as a collection of separate neurites, which gets the specific feature for a neuron object.

My rationale when I created the two feature categories was to have the ability to apply some features in both resolution orders, if it is required.

**Comment by wizmer** [comment]
Is this PR still relevant ? Should we merge it ?

**Comment by mgeplf** [comment]
Good question, I don't really know.  I think some of the changes are useful, but the decorator to register features is probably not a good idea.  I also think we could drop the 'register your own features', b/c I doubt anyone has used that.

This was one of the main reasons for doing it:
https://github.com/BlueBrain/NeuroM/pull/572/files#r93192751

**Comment by asanin-epfl** [comment]
closing in favor of #860
